tempesta_fw/http_parser.c

```diff
@@ -1370,6 +1393,332 @@ __parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len)
 	return r;
 }
 
+#define GROW_STR(str, data, sz)						\
+do {									\
+	TfwStr *last_chunk = TFW_STR_LAST(str);				\
+	if ((str)->len > 0 &&						\
+	    (char *)(str)->ptr + (str)->len == (char *)(data))		\
+	{								\
+		if (last_chunk != (str))				\
+			last_chunk->len += (sz);			\
+		(str)->len += (sz);					\
+	} else {							\
+		__msg_field_fixup_pos(str, data, sz);			\
+	}								\
+} while (0)
+
+/*
+ * Resolves all escaped characters, removes heading and trailing double quotes.
+ * Calling site should ensure dst->ptr have enough writable memory. Resulting
+ * string is at most src->len bytes long.
+ */
+static void
+unescape_multipart_boundary(TfwStr * __restrict dst, TfwStr * __restrict src)
+{
+	TfwStr *c, *c_end;
+	bool esc = false;
+	char *pos;
+	char *last;
+
+	/* There are at least two double quotes. */
+	BUG_ON(src->len < 2);
+
+	if (TFW_STR_PLAIN(src)) {
+		c = src;
+		c_end = src + 1;
+	} else {
+		c = src->ptr;
+		c_end = c + TFW_STR_CHUNKN(src);
+	}
+
+	BUG_ON(c->len == 0);
+	pos = (char *)c->ptr + 1;
+	last = (char *)c->ptr + c->len;
+	dst->len = 0;
+	while (1) {
+		if (pos == last) {
+			c += 1;
+			if (c == c_end)
+				break;
+			pos = c->ptr;
+			last = pos + c->len;
+		}
+
+		if (esc) {
+			((char *)dst->ptr)[dst->len++] = *pos++;
+			esc = false;
+			continue;
+		}
+
+		if (unlikely(*pos == '"'))
+			return;
+
+		if (*pos == '\\') {
+			esc = true;
+			pos += 1;
+			continue;
+		}
+
+		((char *)dst->ptr)[dst->len++] = *pos++;
+	}
+}
+
+static int
+__req_parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	TfwHttpReq *req = (TfwHttpReq *)hm;
+	__FSM_DECLARE_VARS(hm);
+
+	__FSM_START(parser->_i_st) {
+
+	__FSM_STATE(I_ContType) {
+		if (req->method != TFW_HTTP_METH_POST)
+			__FSM_I_MOVE_n(I_EoL, 0);
+		__FSM_I_MOVE_n(I_ContTypeMediaType, 0);
+	}
+
+	__FSM_STATE(I_ContTypeMediaType) {
+		TRY_STR_LAMBDA("multipart/form-data", {},
+		               I_ContTypeMaybeMultipart);
+		TRY_STR_INIT();
+		__FSM_I_MOVE_n(I_ContTypeOtherType, 0);
+	}
+
+	__FSM_STATE(I_ContTypeMaybeMultipart) {
+		switch (*p) {
+		case ';':
+			__FSM_I_MOVE(I_ContTypeMultipart);
+		case ' ':
+		case '\t':
+			__FSM_I_MOVE(I_ContTypeMultipartOWS);
```

Comment:
Yes, that's exactly why this state is needed. However, I think it will be needed anyway, as a fallback for the case there received buffer ends exactly at `/form-data` and we have no data beyond that at the moment of parsing. There are similar cases in the main parser where field names are matched: fast and slow paths.